File: BusCatalog.Api/BusCatalog.Api/Domain/HealthCheck/HealthCheckService.cs
using System.Net;
using BusCatalog.Api.Domain.HealthCheck.Ports;
using BusCatalog.Api.Infrastructure.Configurations;

namespace BusCatalog.Api.Domain.HealthCheck;

public interface IHealthCheckService
{
    HealthCheckResponse SelfHealthCheck();
    Task<HealthCheckResponse> SpaHealthCheckAsync();
    Task<HealthCheckResponse> ScraperHealthCheckAsync();
}

public sealed class HealthCheckService : IHealthCheckService
{
    private readonly IHealthCheckAdapter _adapter;
    private readonly IHttpContextAccessor _httpContextAccessor;
    private readonly IConfiguration _configuration;

    public HealthCheckService(
        IHealthCheckAdapter adapter,
        IHttpContextAccessor httpContextAccessor,
        IConfiguration configuration)
    {
        _adapter = adapter;
        _httpContextAccessor = httpContextAccessor;
        _configuration = configuration;
    }

    public HealthCheckResponse SelfHealthCheck() =>
        HealthCheckResponse.Generate(
            _httpContextAccessor.HttpContext!.GetFullRequestUrl(),
            (int)HttpStatusCode.OK,
            success: true);

    public Task<HealthCheckResponse> SpaHealthCheckAsync() =>
        _adapter.SendAsync(
            _configuration[nameof(ConfigurationKeys.SpaUrl)]!,
            nameof(HealthCheckApplication.Spa));

    public Task<HealthCheckResponse> ScraperHealthCheckAsync() =>
        _adapter.SendAsync(
            $"{_configuration[nameof(ConfigurationKeys.ScraperUrl)]}/api/health-check",
            nameof(HealthCheckApplication.Scraper));
}

File: BusCatalog.Scraper/src/main/java/br/com/marcomvidal/buscatalog/scraper/healthcheck/HealthCheckService.java
package br.com.marcomvidal.buscatalog.scraper.healthcheck;

import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

import br.com.marcomvidal.buscatalog.scraper.healthcheck.ports.HealthCheckResponse;
import br.com.marcomvidal.buscatalog.scraper.infrastructure.configuration.sections.ConfigurationSection;
import jakarta.servlet.http.HttpServletRequest;

@Service
public class HealthCheckService {
    private final HealthCheckAdapter adapter;
    private HttpServletRequest request;
    private final ConfigurationSection configuration;
    private final String API_HEALTH_CHECK_URL = "/api/health-check";

    public HealthCheckService(
        HealthCheckAdapter adapter,
        HttpServletRequest request,
        ConfigurationSection configuration) {
        this.adapter = adapter;
        this.request = request;
        this.configuration = configuration;
    }

    public HealthCheckResponse selfHealthCheck() {
        return HealthCheckResponse.generate(
            request.getRequestURL().toString(),
            HttpStatus.OK.value(),
            true);
    }

    public HealthCheckResponse apiHealthCheck() {
        return adapter.send(
            configuration.getApiUrl() + API_HEALTH_CHECK_URL,
            HealthCheckApplication.API.name());
    }
}
